text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,249 @@
+   by nature, so wouldn't need `mut` under that definition.
+ - If you can't get a `&mut` to a local declared without `mut`, how does `drop` do it?
+ - There is also the "move loophole," where moving out of a binding lets you mutate it:
+   ```rs
+   let not_mutable = vec![1, 2, 3];
+   let mut not_mutable = not_mutable;
+   not_mutable.push(4);
+   ```
+ - But that loophole _doesn't_ work for `move` closures, which ostensibly do the same thing:
+   ```rs
+   let not_mutable = vec![1, 2, 3];
+   (move || { 
+      not_mutable.push(4);
+   })();
+   ```
+
+The IRLO and Zulip threads include more examples of inconsistency, and also some examples of people
+being confused by the inconsistency or mistakenly believing that the `mut_non_mut` check
+is related to Rust's static safety guarantees.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+> Explain the proposal as if it was already included in the language and you were teaching
+> it to another Rust programmer.
+
+In addition to using `&mut` to declare an exclusive reference, the `mut` keyword may be used to indicate
+that a local variable or pattern binding is mutated later on.  The `unused_mut` and
+`mut_non_mut` lints will notify you when a `mut` binding is never mutated or when a non-`mut` binding is
+mutated, respectively.  Ensuring correct usage of `mut` annotations on all local variables will help
+future readers of your code to better understand which values are and are not expected to change.
+
+The below will warn that you added a `mut` annotation to a local that is not mutated:
+```rs
+fn get_num() -> i32 {
+    let mut x = 12; // WARNING! Variable does not need to be mutable
+    x
+}
+```
+
+The below will warn that you omitted the `mut` annotation from a local that was re-assigned:
+```rs
+fn get_num() -> i32 {
+    let x = 12;
+    x = 13; // WARNING! Assigned twice to immutable variable `x`
+    x
+}
+```
+
+Creating a `&mut` reference is also considered a mutation, since a mutation could occur through the reference:
+```rs
+fn get_num() -> i32 {
+    let x = 12;
+    let px = &mut x; // WARNING! Borrowing `x` as mutable when it is not declared as mutable
+    *px = 13;
+    x
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The borrow checker has to deal with the `mut` vs `&mut` distinction, and therefore
+already has special handling for locals.  Making immutable re-assignment a lint just
+means converting local-specific code paths in borrowck to trigger lints instead of fatal
+diagnostics.  Because this is a relaxation of existing language semantics,
+the change is fully backwards-compatible, and making the lint deny-by-default means even code which _fails_
+to compile will continue to do so after the change. 
+
+One interesting consequence of this change is that it allows mutation in places where there
+previously was no syntax to allow it, for example in `ref` bindings:
```

Comment:
Syntax for this exists behind `#![feature(mut_ref)]`: `mut ref mut` and `mut ref`.

Reply:
Hm, ok.  I actually don't know of any other examples off the top of my head where `mut` syntax is missing, so maybe the whole "syntax doesn't exist" point is actually moot.  I've added a note.